File: src/main/java/day7/Day7.java
package day7;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.function.BiFunction;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class Day7 {
    public static void main(String[] args) {
        List<Equation> equations;
        try (var lines = Files.lines(new File("./input").toPath())){
            equations = lines
                    .map(Day7::parseEquation)
                    .toList();
        } catch (IOException error) {
            throw new RuntimeException(error);
        }

        long resultA = 0L;
        List<Operator> operatorsForA = Arrays.asList(Operator.values()).subList(0, 2);
        for (Equation equation : equations) {
            if (possible(equation, operatorsForA)) {
                resultA += equation.result();
            }
        }
        System.out.println("A: " + resultA);

        long resultB = 0L;
        for (Equation equation : equations) {
            if (possible(equation, Arrays.asList(Operator.values()))) {
                resultB += equation.result();
            }
        }
        System.out.println("B: " + resultB);
    }

    record Equation(long result, List<Long> input) {}

    static Equation parseEquation(String line) {
        String[] parts = line.split(": ");
        long result = Long.parseLong(parts[0]);
        List<Long> input = Stream.of(parts[1].split(" "))
                .map(Long::parseLong)
                .collect(Collectors.toList());
        return new Equation(result, input);
    }

    enum Operator {
        Addition(Long::sum),
        Multiplication((lhs, rhs) -> lhs * rhs),
        Concatenation((lhs, rhs) -> Long.parseLong(Long.toString(lhs) + rhs));

        private final BiFunction<Long, Long, Long> implementation;

        Operator(BiFunction<Long, Long, Long> implementation) {
            this.implementation = implementation;
        }

        public long apply(long lhs, long rhs) {
            return implementation.apply(lhs, rhs);
        }
    }

    static class OperatorChains implements Iterator<List<Operator>> {
        final List<Operator> availableOperators;
        final int length;
        final int period;
        int state = 0;

        public OperatorChains(List<Operator> availableOperators, int length) {
            this.availableOperators = availableOperators;
            this.length = length;
            this.period = (int) Math.pow(availableOperators.size(), length);
        }

        @Override
        public boolean hasNext() {
            return state < period;
        }

        @Override
        public List<Operator> next() {
            List<Operator> result = new ArrayList<>();
            int chunk = state;
            for (int chunkIndex = 0; chunkIndex < length; ++chunkIndex) {
                int operatorIndex = chunk % availableOperators.size();
                chunk /= availableOperators.size();
                result.add(availableOperators.get(operatorIndex));
            }
            ++state;
            return result;
        }
    }

    static boolean possible(Equation equation, List<Operator> operators) {
        for (OperatorChains chain = new OperatorChains(operators, equation.input().size() - 1); chain.hasNext(); ) {
            if (correct(equation, chain.next())) {
                return true;
            }
        }
        return false;
    }

    static boolean correct(Equation equation, List<Operator> chain) {
        long result = equation.input().get(0);
        for (int index = 0; index < equation.input().size() - 1; ++index) {
            Operator operator = chain.get(index);
            long nextInputValue = equation.input().get(index + 1);
            result = operator.apply(result, nextInputValue);
        }
        return result == equation.result();
    }
}

File: src/main/kotlin/day7/day7.kt
package day7

import java.io.File
import kotlin.math.pow
import kotlin.math.roundToInt

fun main() {
    val equations = File("input")
        .readLines()
        .map(::parseEquation)

    var resultA = 0L
    val operatorsForA = Operator.entries.subList(0, 2)
    for (equation in equations) {
        if (possible(equation, operatorsForA)) {
            resultA += equation.result
        }
    }
    println("A: $resultA")

    var resultB = 0L
    for (equation in equations) {
        if (possible(equation, Operator.entries)) {
            resultB += equation.result
        }
    }
    println("B: $resultB")
}

data class Equation(val result: Long, val input: List<Long>)

fun parseEquation(line: String): Equation {
    val (resultText, inputText) = line.split(": ")
    val result = resultText.toLong()
    val input = inputText.split(" ")
        .map({ it.toLong() })
    return Equation(result, input)
}

enum class Operator(val apply: (Long,Long) -> Long) {
    Addition({ lhs, rhs -> lhs + rhs }),
    Multiplication({ lhs, rhs -> lhs * rhs}),
    Concatenation({ lhs, rhs -> (lhs.toString() + rhs.toString()).toLong() })
}

class operatorChains(val availableOperators: List<Operator>, val length: Int): Iterator<List<Operator>> {
    val period = availableOperators.size.toDouble().pow(length.toDouble()).roundToInt()
    var state = 0

    override fun hasNext(): Boolean {
        return state < period
    }

    override fun next(): List<Operator> {
        val result = mutableListOf<Operator>()
        var chunk = state
        for (chunkIndex in 0..<length) {
            val operatorIndex = chunk % availableOperators.size
            chunk /= availableOperators.size
            result.add(availableOperators[operatorIndex])
        }
        ++state
        return result
    }
}

fun possible(equation: Equation, operators: List<Operator>): Boolean {
    for (chain in operatorChains(operators, equation.input.size - 1)) {
        if (correct(equation, chain)) {
            return true
        }
    }
    return false
}

fun correct(equation: Equation, chain: List<Operator>): Boolean {
    var result = equation.input[0]
    for (index in 0..<equation.input.size - 1) {
        val operator = chain[index]
        val nextInputValue = equation.input[index + 1]
        result = operator.apply(result, nextInputValue)
    }
    return result == equation.result
}
